ft-storage/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+#![no_std]
+use ft_storage_io::*;
+use gstd::{msg, prelude::*, ActorId};
+use primitive_types::H256;
+
+#[derive(Default)]
+struct FTStorage {
+    ft_logic_id: ActorId,
+    transaction_status: BTreeMap<H256, bool>,
+    balances: BTreeMap<ActorId, u128>,
+    approvals: BTreeMap<ActorId, BTreeMap<ActorId, u128>>,
+}
+
+static mut FT_STORAGE: Option<FTStorage> = None;
+
+impl FTStorage {
+    fn get_balance(&self, account: &ActorId) {
+        let balance = self.balances.get(account).unwrap_or(&0);
+        msg::reply(FTStorageEvent::Balance(*balance), 0).expect("");
+    }
+
+    fn increase_balance(&mut self, transaction_hash: H256, account: &ActorId, amount: u128) {
+        self.assert_ft_contract();
+
+        // check transaction status
+        if let Some(status) = self.transaction_status.get(&transaction_hash) {
+            match status {
+                true => reply_ok(),
+                false => reply_err(),
+            };
+            return;
+        }
+
+        // increase balance
+        self.balances
+            .entry(*account)
+            .and_modify(|balance| *balance = (*balance).saturating_add(amount))
+            .or_insert(amount);
+
+        self.transaction_status.insert(transaction_hash, true);
+        msg::reply(FTStorageEvent::Ok, 0).expect("error in sending a reply `FTStorageEvent::Ok");
+    }
+
+    fn decrease_balance(
+        &mut self,
+        transaction_hash: H256,
+        msg_source: &ActorId,
+        account: &ActorId,
+        amount: u128,
+    ) {
+        self.assert_ft_contract();
+        // check transaction status
+        if let Some(status) = self.transaction_status.get(&transaction_hash) {
+            match status {
+                true => reply_ok(),
+                false => reply_err(),
+            };
+            return;
+        }
+
+        // decrease balance
+        if let Some(balance) = self.balances.get_mut(account) {
+            if *balance >= amount {
+                if msg_source == account {
+                    *balance -= amount;
+                    self.transaction_status.insert(transaction_hash, true);
+                    reply_ok();
+                    return;
+                } else if let Some(allowed_amount) = self
+                    .approvals
+                    .get_mut(account)
+                    .and_then(|m| m.get_mut(msg_source))
+                {
+                    *balance -= amount;
+                    *allowed_amount -= amount;
+                    self.transaction_status.insert(transaction_hash, true);
+                    reply_ok();
+                    return;
```

Comment:
```suggestion
                    if *allowed_amount >= amount {
                        *balance -= amount;
                        *allowed_amount -= amount;
                        self.transaction_status.insert(transaction_hash, true);
                        reply_ok();
                        return;
                    }
```

and I think you should simplify this complex expression

Reply:
That's true, already fixed it